contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -3093,3 +3093,116 @@ is_valid_func_ownership_chain(void *expr, Oid objectOwnerId)
+        return;
+    
+    /* For OBJECT_TYPE, convert List to TypeName if needed */
+    if (stmt->objectType == OBJECT_TYPE && IsA(stmt->object, List))
+    {
+        TypeName *typename = makeTypeNameFromNameList((List *) stmt->object);
+        object = (Node *) typename;
+    }
+    
+    /* Get object address to determine schema */
+    address = get_object_address(stmt->objectType, object, &relation, AccessShareLock, false);
+    
+    if (stmt->objectType == OBJECT_SCHEMA)
+    {
+		/* For schema objects, the object itself is the schema */
+        schema_oid = address.objectId;
+    }
+    else
+    {
+		/* For other objects, get their containing schema */
+        schema_oid = get_object_namespace(&address);
+    }
+
+	if (!OidIsValid(schema_oid))
+	{
+		if (relation)
+			RelationClose(relation);
+		return;
+	}
+    
+    schema_name = get_namespace_name(schema_oid);
+    if (relation)
+        RelationClose(relation);
+    
+    if (!schema_name)
+        return;
+
+    /* Check if it's a Babelfish schema */
+    if (physical_schema_name_exists(schema_name))
+    {
+        pfree(schema_name);
+        ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                 errmsg("ALTER .. OWNER .. is blocked in PG dialect on TSQL objects. Please set babelfishpg_tsql.enable_alter_owner_from_pg to true to enable.")));
+    }
+
+    if (schema_name)
+        pfree(schema_name);
+}
+
+/*
+ * restrict_alter_table_stmt  
+ * Blocks ALTER TABLE statements from PG dialect on TSQL objects.
+ * Extracts schema name from the table relation and checks if it belongs to Babelfish.
+ */
+void
+restrict_alter_table_stmt(AlterTableStmt *stmt)
+{
+    char *schema_name = NULL;
+    Oid schema_oid = InvalidOid;
+    ListCell *lcmd;
+
+    /* Skip if this is ALTER VIEW - handled with babelfishpg_tsql.enable_create_alter_view_from_pg */
+    if (stmt->objtype == OBJECT_VIEW)
+        return;
+
+    /* Extract schema from RangeVar */
+    if (stmt->relation)
+    {
+        if (stmt->relation->schemaname)
+            schema_name = stmt->relation->schemaname;
+        else
+        {
+            schema_oid = RangeVarGetRelid(stmt->relation, AccessExclusiveLock, true);
+            if (OidIsValid(schema_oid))
+                schema_name = get_namespace_name(schema_oid);
+        }
+    }
+
+    /* Check if any command is AT_ChangeOwner */
+    foreach(lcmd, stmt->cmds)
+    {
+        AlterTableCmd *cmd = (AlterTableCmd *) lfirst(lcmd);
+        if (cmd->subtype != AT_ChangeOwner)
+            continue;
+
+        /* Check if it's a Babelfish schema */
+        if (schema_name && physical_schema_name_exists(schema_name))
+        {
+            ereport(ERROR,
+                    (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                     errmsg("ALTER .. OWNER .. is blocked in PG dialect on TSQL objects. Please set babelfishpg_tsql.enable_alter_owner_from_pg to true to enable.")));
+        }
+        break;
```

Comment:
I think we shouldn't break here since we want to traverse through all the AT_ChangeOwner commands and not just the first one.